packages/metadata/src/decorator-factory.ts

```diff
@@ -0,0 +1,539 @@
+   */
+  getTarget(spec: T) {
+    if (typeof spec === 'object' && spec != null) {
+      const specWithTarget = spec as any;
+      return specWithTarget[DecoratorFactory.TARGET];
+    }
+  }
+
+  /**
+   * To be overridden by subclasses to process inherited metadata
+   * @param meta Metadata inherited from the base
+   * @param target Decoration target
+   * @param member Optional property or method
+   * @param descriptorOrIndex Optional parameter index or method descriptor
+   */
+  protected processInherited(
+    baseMeta: M,
+    target: Object,
+    member?: string | symbol,
+    descriptorOrIndex?: TypedPropertyDescriptor<any> | number,
+  ): M {
+    return baseMeta;
+  }
+
+  /**
+   * To be overridden by subclasses to process local metadata
+   * @param meta Metadata exists on the target
+   * @param target Decoration target
+   * @param member Optional property or method
+   * @param descriptorOrIndex Optional parameter index or method descriptor
+   */
+  protected processLocal(
+    localMeta: M,
+    target: Object,
+    member?: string | symbol,
+    descriptorOrIndex?: TypedPropertyDescriptor<any> | number,
+  ): M {
+    return localMeta;
+  }
+
+  /**
+   * Create a decorator function of the given type. Each sub class MUST
+   * implement this method.
+   */
+  create(): D {
+    throw new Error('create() is not implemented');
+  }
+
+  /**
+   * Base implementation of the decorator function
+   * @param target Decorator target
+   * @param member Optional property or method
+   * @param descriptorOrIndex Optional method descriptor or parameter index
+   */
+  protected decorate(
+    target: Object,
+    member?: string | symbol,
+    descriptorOrIndex?: TypedPropertyDescriptor<any> | number,
+  ) {
+    const targetName = this.getTargetName(target, member, descriptorOrIndex);
+    let meta: M = Reflector.getOwnMetadata(this.key, target);
+    if (meta == null && this.allowInheritance) {
+      // Clone the base metadata so that it won't be accidentally
+      // mutated by sub classes
+      meta = cloneDeep(Reflector.getMetadata(this.key, target));
+      meta = this.processInherited(meta, target, member, descriptorOrIndex);
+      if (debug.enabled) {
+        debug('%s: %j', targetName, meta);
+      }
+      Reflector.defineMetadata(this.key, meta, target);
+    } else {
+      meta = this.processLocal(meta, target, member, descriptorOrIndex);
+      if (debug.enabled) {
+        debug('%s: %j', targetName, meta);
+      }
+      Reflector.defineMetadata(this.key, meta, target);
+    }
+  }
+
+  /**
+   * Create a decorator function
+   * @param key Metadata key
+   * @param spec Metadata object from the decorator function
+   * @param allowInheritance Inheritance will be honored, default to `true`
+   */
+  protected static _createDecorator<
+    T,
+    M extends T | MetadataMap<T> | MetadataMap<T[]>,
+    D extends DecoratorType
+  >(key: string, spec: T, allowInheritance: boolean = true): D {
```

Comment:
nitpick: 
> `allowInheritance: boolean = true` 

not sure will we have other options like `allowInheritance` in the future, how about making the third parameter an option object?

Reply:
It sounds good to me.